customers.go

```diff
@@ -61,3 +67,121 @@ func BuildCustomerDTO(customer shopify.Customer) CustomerDTO {
 
 	return customerDTO
 }
+
+type customerRepository struct {
+	client    http.Client
+	createURL func(endpoint string) string
+}
+
+func newCustomerRepository(client http.Client, createURL func(endpoint string) string) customerRepository {
+	return customerRepository{
+		client,
+		createURL,
+	}
+}
+
+func (c customerRepository) Update(customer shopify.Customer) (shopify.Customer, error) {
+	// Map to DTO
+	customerDTO := BuildCustomerDTO(customer)
+
+	request := struct {
+		CustomerDTO `json:"customer"`
+	}{
+		customerDTO,
+	}
+
+	body, err := json.Marshal(request)
+	if err != nil {
+		return shopify.Customer{}, err
+	}
+
+	url := c.createURL(fmt.Sprintf("customers/%d.json", customer.ID))
+
+	respBody, _, err := c.client.Put(url, body, nil)
+	if err != nil {
+		switch err.(type) {
+		case http.ErrHTTP:
+			var responseError ResponseError
+			e := json.Unmarshal([]byte(err.(http.ErrHTTP).Body), &responseError)
+			if e != nil {
+				return shopify.Customer{}, err
+			}
+
+			return shopify.Customer{}, responseError.getType(err)
+		default:
+			return shopify.Customer{}, err
+		}
+	}
+
+	var response struct {
+		Customer CustomerDTO `json:"customer"`
+	}
+
+	err = json.Unmarshal(respBody, &response)
+	if err != nil {
+		return shopify.Customer{}, err
+	}
+
+	return response.Customer.ToShopify(), nil
+}
+
+// ResponseError is used to store the error response for a customer
+type ResponseError struct {
+	Errors Fields `json:"errors"`
+}
+
+// Fields holds all the possible fields for an error
+type Fields struct {
+	Email []interface{} `json:"email"`
+	Phone []interface{} `json:"phone"`
+}
+
+func (e ResponseError) getType(err error) error {
+	if len(e.Errors.Email) > 0 {
+		return ErrCustomerEmail{
+			Msg: err.Error(),
+			Err: err,
+		}
+	}
+
+	if len(e.Errors.Phone) > 0 {
+		return ErrCustomerPhone{
+			Msg: err.Error(),
+			Err: err,
+		}
+	}
+
+	return nil
+}
+
+// ErrCustomerEmail is thrown when there has been an error with a customers email
+type ErrCustomerEmail struct {
+	Msg string
+	Err error
+}
```

Comment:
What if we have only one struct that holds `Msg` and `Err` instead of two and we create email and phone errors off that?

Something like:


`
type errCustomer struct {
	Msg string
	Err error
}
`

`
EmailError errCustomer
....
`

`
PhoneError errCustomer
.....
`

Reply:
Well because it defines one interface it expects it to be one error really.